rllib/agents/dqn/distributional_q_model.py

```diff
@@ -151,14 +151,33 @@ def build_state_score(model_out):
                     state_out, units=num_atoms, activation=None)
             return state_score
 
-        def build_action_value_in_scope(model_out):
-            with tf.variable_scope(
-                    name + "/action_value", reuse=tf.AUTO_REUSE):
-                return build_action_value(model_out)
-
-        def build_state_score_in_scope(model_out):
-            with tf.variable_scope(name + "/state_value", reuse=tf.AUTO_REUSE):
-                return build_state_score(model_out)
+        if tf.executing_eagerly():
+            # have to use a variable store to reuse variables in eager mode
+            import tensorflow.contrib as tfc
+            store = tfc.eager.EagerVariableStore()
```

Comment:
Somehow DQN still doesn't work in eager mode, due to most of the variables (10/14) having a None gradient. This doesn't seem to be an issue for SimpleQ, so I assume the difficulty is something to do with the distributionalqmodel (though it could also be something else in the policy).

Reply:
Have you tried with something like a simple hardcoded keras layers that doesn't require using variable scopes? Also, what variables are connected (non-none grad) might give us clue about where things break. Would it possible to print out the name of the variables?

Reply:
Figured it out -- the variable scope the function was executing in was changing, so a duplicate set of variable was created.